### PR TITLE
[Python 2 to 3] Main core fixes + regression.uts

### PR DESCRIPTION
**Pull request by gpotter2**
**Python 3 support major PR. It's big, but most of it are regression.uts tests fixes.**

- Requires https://github.com/secdev/scapy/pull/674
- Scapy boots and most of its major functions work 

This PR:
- fixes a lot of ord/orb, bytes chr/chb, compat bugs
- fixes a few .next() / next()
- fixes most layers
- the fixes are mostly in `regression.uts`

Current status (updated):
- regression.uts: 1069/1091 tests passing

### CODE REVIEW

File: scapy/compat.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
+## This file is part of Scapy
+## See http://www.secdev.org/projects/scapy for more informations
+## Copyright (C) Philippe Biondi <phil@secdev.org>
```
Comment (guedou) [inline scapy/compat.py]: I think that it is fair to put your name and address here =)

File: scapy/compat.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+from __future__ import absolute_import
+import codecs
+
+import scapy.modules.six as six
+
+###########
+# Python3 #
+###########
+
+def cmp_to_key(mycmp):
```
Comment (guedou) [inline scapy/compat.py]: Could you write an example in the comment ?

File: scapy/compat.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        def __ne__(self, other):
+            return mycmp(self.obj, other.obj) != 0
+    return K
+
+def cmp(a, b):
+    """Old python 2 function"""
+    return (a > b) - (a < b)
+
+def orb(x):
+    """Same than orb() but handle python 3 too
```
Comment (guedou) [inline scapy/compat.py]: s/python/Python/ everywhere

File: scapy/compat.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                return bytes([x])
+            except ValueError:
+                # Ignore out of range strings
+                return bytes(str(x), "utf8")
+        if isinstance(x, bytes):
+            return x
+        if hasattr(x, "__bytes__"):
+            return bytes(x)
+        return bytes(x, "utf8")
+    def plain_str(x):
```
Comment (guedou) [inline scapy/compat.py]: add \n before the function definition (+ same comment below)

File: scapy/config.py [hunk 5]
```diff
@@ -171,14 +171,17 @@ def register(self, cmd):
 def lsc():
     print(repr(conf.commands))
 
-class CacheInstance(dict):
+class CacheInstance(dict, object):
```
Comment (guedou) [inline scapy/config.py]: Shall this go in a separated PR ?

File: scapy/contrib/bgp.uts [hunk 6]
```diff
@@ -297,7 +297,7 @@ str(BGPOptParam()) == b'\x02\x00'
 
 = BGPOptParam - Instantiation with specific values (1)
 str(BGPOptParam(param_type = 1)) == b'\x01\x00'
-str(BGPOptParam(param_type = 1, param_value = BGPAuthenticationInformation())) == b'\x01\x04'
+str(BGPOptParam(param_type = 1, param_value = BGPAuthenticationInformation())) == b'\x01\x00'
```
Comment (guedou) [inline scapy/contrib/bgp.uts]: Why is this needed ?

File: scapy/fields.py [hunk 7]
```diff
@@ -52,6 +53,8 @@ def i2count(self, pkt, x):
         return 1
     def h2i(self, pkt, x):
         """Convert human value to internal value"""
+        if type(x) is str:
+            x = raw([ord(i) for i in x])
```
Comment (guedou) [inline scapy/fields.py]: s/ord/orB/ ?

File: scapy/fields.py [hunk 8]
```diff
@@ -556,6 +563,9 @@ def getfield(self, pkt, s):
         l = self.length_from(pkt)
         return s[l:], self.m2i(pkt,s[:l])
 
+def _toHex(x):
```
Comment (guedou) [inline scapy/fields.py]: rename it to `to_hex()` to conform with other Scapy function name. This could go in utils.py

File: scapy/layers/tls/cert.py [hunk 9]
```diff
@@ -128,32 +131,33 @@ def __call__(cls, obj_path, obj_max_size, pem_marker=None):
...
+        obj_path = raw(obj_path)
 
         if (not b'\x00' in obj_path) and os.path.isfile(obj_path):
             _size = os.path.getsize(obj_path)
             if _size > obj_max_size:
                 raise Exception(error_msg)
             try:
                 f = open(obj_path)
-                raw = f.read()
+                raw_ = f.read()
```
Comment (guedou) [inline scapy/layers/tls/cert.py]: s/raw_/_raw/ do be consistent with the changes you made to the variables called bytes ?

File: scapy/layers/tls/session.py [hunk 10]
```diff
@@ -963,7 +963,7 @@ def __repr__(self):
                 if len(sid) > 12:
                     sid = sid[:11] + "..."
                 res.append((src, dst, sid))
-        colwidth = map(lambda x: max(map(lambda y: len(y), x)), apply(zip, res))
+        colwidth = [max([len(y) for y in x]) for x in zip(*res)]
```
Comment (guedou) [inline scapy/layers/tls/session.py]: Use a generator () instead of []

File: scapy/tools/UTscapy.py [hunk 11]
```diff
@@ -301,12 +302,16 @@ def dump_campaign(test_campaign):
...
-
-def crc32(x):
-    return "%08X" % (0xffffffff & zlib.crc32(x))
-
-def sha1(x):
-    return sha.sha(x).hexdigest().upper()
+if six.PY2:
+    def crc32(x):
+        return "%08X" % (0xffffffff & zlib.crc32(x))
+    def sha1(x):
```
Comment (guedou) [inline scapy/tools/UTscapy.py]: Add  \n before the definition (+ same comment below)

File: scapy/arch/linux.py [hunk 12]
```diff
@@ -146,17 +149,20 @@ def attach_filter(s, bpf_filter, iface):
...
     nb = int(lines[0])
-    bpf = ""
+    bpf = b""
     for l in lines[1:]:
         bpf += struct.pack("HBBI", *(int(e) for e in l.split()))
 
-    # XXX. Argl! We need to give the kernel a pointer on the BPF,
-    # python object header seems to be 20 bytes. 36 bytes for x86 64bits arch.
-    bpfh = struct.pack("HL", nb, id(bpf) + (36 if IS_64BITS else 20))
+    # XXX. Argl! We need to give the kernel a pointer on the BPF
```
Comment (gpotter2) [inline scapy/arch/linux.py]: passing the pointer through `id()` isn't accepted anymore on python 3, or at least we can't make it python 2 compatible

File: scapy/asn1/mib.py [hunk 13]
```diff
@@ -15,43 +15,46 @@
...
 from scapy.utils import do_graph
 import scapy.modules.six as six
+from scapy.compat import *
 
 #################
 ## MIB parsing ##
 #################
 
-_mib_re_integer = re.compile("^[0-9]+$")
+_mib_re_integer = re.compile(b"^[0-9]+$")
```
Comment (gpotter2) [inline scapy/asn1/mib.py]: OIDS are used as bytes, whereas others are used as strings

File: scapy/config.py (hunk 5 above)
Comment (gpotter2) [inline scapy/config.py]: Surely

File: scapy/contrib/bgp.uts (hunk 6 above)
Comment (gpotter2) [inline scapy/contrib/bgp.uts]: Explanation: BGP was using i2m method that did not check if the field was empty. Here: "\x04" stands for `len(...) = 4` because the string "None" was passed to the function. It's a bug fixed indirectly

**Comment by guedou** [comment]
@gpotter2 shall we start reviewing this PR ?

**Comment by gpotter2** [comment]
@guedou Thanks, but i need to rebase every thing against https://github.com/secdev/scapy/pull/674, so it needs to be merged first

**Comment by p-l-** [comment]
@gpotter2 #674 has been merged, let us know when you need us on that one.

**Comment by gpotter2** [comment]
@guedou @p-l- Should be ready for reviewing.
I have a bug on OSX only with the temp files that i don't understand... Any ideas ?

File: run_scapy3.bat [hunk 14]
```diff
@@ -1,7 +1,7 @@
 @echo off
 set PYTHONPATH=%cd%
 set PYTHONDONTWRITEBYTECODE=True
-"C:\Program Files (x86)\Python3\python.exe" -m scapy.__init__ %*
+"D:\Programms\Python3\python.exe" -m scapy.__init__ %*
```
Comment (p-l-) [inline run_scapy3.bat]: Do you really want to commit this change?

Comment (gpotter2) [reply run_scapy3.bat]: Not really..., i will remove this on the next patch of this PR.

**Comment by gpotter2** [comment]
Alright, this PR starts to be pretty big. I'll stop it now.

Currently, scapy can send/recieve packets, and perform basic tasks.
regression.uts: 932/1090 tests passing

File: scapy/arch/common.py [hunk 15]
```diff
@@ -18,3 +20,35 @@ def get_if(iff, cmd):
...
+
+def get_bpf_pointer(tcpdump_lines):
+    """Create a BPF Pointer for TCPDump filter"""
+    # Allocate BPF instructions
+    size = int(tcpdump_lines[0])
+    bpf_insn_a = bpf_insn * size
+    bip = bpf_insn_a()
+
+    # Fill the BPF instruction structures with the byte code
+    lines = lines[1:]
```
Comment (p-l-) [inline scapy/arch/common.py]: Do you mean `tcpdump_lines`?

File: scapy/arch/common.py [hunk 16]
```diff
@@ -18,3 +20,35 @@ def get_if(iff, cmd):
...
+    for i in range(len(tcpdump_lines)):
```
Comment (p-l-) [inline scapy/arch/common.py]: Why don't you use `for line in tcdpump_lines`?

**Comment by gpotter2** [comment]
@p-l- Fixed, thanks !
I'm not very used to the `attach_filter` function :/ Any ideas of what's happening ? I can't test on linux right now :/

File: scapy/arch/linux.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/arch/linux.py]: You're right also

**Comment by gpotter2** [comment]
Ready to be reviewed. Tests passing

**Comment by ebraminio** [comment]
@gpotter2: As an outsider and not also very familiar with the project, I feel I can still thank you for the great work you've done here, keep up the good work :)

File: test/pipetool.uts [hunk 17]
```diff
@@ -263,6 +263,7 @@ os.unlink("t2.pcap")
 
 = Test InjectSink and Inject3Sink
 ~ needs_root
+import mock
```
Comment (p-l-) [inline test/pipetool.uts]: This is not needed (see next line)

Comment (gpotter2) [reply test/pipetool.uts]: Very right

**Comment by gpotter2** [comment]
@ebraminio Thanks for your support :)

**Comment by gpotter2** [comment]
@guedou @p-l- I won't add anything to this PR, should be ready for reviewing ^^

**Comment by p-l-** [comment]
OK, let's merge this one, and fix what we need to fix ;-)
